weavertest/init.go

```diff
@@ -28,48 +28,57 @@ import (
 )
 
-// Mode controls where components live and how their methods are invoked.
-type Mode uint8
-
-const (
-	// Local places all components in the same process and uses local procedure calls
-	// for method invocations.
-	// This is the default value.
-	Local Mode = iota
-
-	// RPC places all components in the same process and uses RPCs for method invocations.
-	RPC
-
-	// Multi places all components in different process (unless explicitly colocated)
-	// and uses RPCs for method invocations on remote components and local procedure calls
-	// for method invocations on colocated components.
-	Multi
+// Runner runs user-supplied testing code as a weaver application.
+type Runner struct {
+	multi    bool // Use multiple processes
+	forceRPC bool // Use RPCs even for local calls
+	name     string
+	config   string
+}
+
+var (
+	// Local is a Runner that places all components in the same process
+	// and uses local procedure calls for method invocations.
+	Local = Runner{name: "Local"}
+
+	// RPC is a Runner that places all components in the same process
+	// and uses RPCs for method invocations.
+	RPC = Runner{multi: false, forceRPC: true, name: "RPC"}
+
+	// Multi is a Runner that places all components in different
+	// process (unless explicitly colocated) and uses RPCs for method
+	// invocations on remote components and local procedure calls for
+	// method invocations on colocated components.
+	Multi = Runner{multi: true, name: "Multi"}
 )
 
-// String returns a string representation of a Mode, suitable for using as sub-test names.
-func (m Mode) String() string {
-	switch m {
-	case RPC:
-		return "RPC"
-	case Local:
-		return "Local"
-	case Multi:
-		return "Multi"
-	default:
-		panic(fmt.Sprintf("unknown mode %d", m))
+// AllRunners returns a slice of all builtin weavertest runners.
+func AllRunners() []Runner { return []Runner{Local, RPC, Multi} }
+
+// WithName returns a new Runner with the specified name. It is useful
+// when the runner has been adjusted and is no longer identical to one
+// of the predefined runners.
+func (r Runner) WithName(name string) Runner { r.name = name; return r }
+
+// WithConfig returns a new Runner with the specified Service Weaver
+// configuration. The config value passed here is identical to what
+// might be found in a Service Weaver config file. It can contain
+// application level as well as component level configuration.
+func (r Runner) WithConfig(config string) Runner { r.config = config; return r }
+
+// Name returns the runner name. It is suitable for use as a sub-test or sub-benchmark name.
+func (r Runner) Name() string {
+	if r.name == "" {
+		return "Default"
 	}
+	return r.name
 }
 
-// AllModes returns a slice of all supported weavertest modes.
-func AllModes() []Mode { return []Mode{Local, RPC, Multi} }
+// SingleProcess returns true iff the runner will run all components in a single process.
+func (r Runner) SingleProcess() bool { return !r.multi }
```

Comment:
I am wondering if it would be clearer for a developer, if we have explicit methods for all the 3 modes instead:

SingleProcessLocalCalls{ ... }
SingleProcessRemoteCalls{ ...}
MultiProcess { ... }

or something like that.

Reply:
At least the callers we have so far are interested in properties that don't exactly match the mode. E.g., SingleProcess() is true for Local+RPC, UsesRPC is true for RPC+Multi.

An alternative would be to just let the caller use "runner == weavertest.Local || runner == weavertest.Multi" etc.?

Reply:
I think it would be cleaner to let the user use the Local/Multi/RPC modes directly. 

I find it confusing that we have to think about single vs multi, but also RPC vs local. 

I don't have strong opinions though, the current implementation is great too.

Reply:
I decided to drop the accesors now and just let the caller compare runner directly. This should suffice since:

1. The need will mostly show up in the weaver implementation since we only use this to test that behavior changes as expected as the runtime changes.
2. The caller can keep info parallel to AllRunners() results if they want (perhaps by avoiding AllRunners() and making their own testCase slice).

Reply:
Sounds good to me.